src/datanode/src/heartbeat/handler/close_region.rs

```diff
@@ -0,0 +1,243 @@
+    ) -> Result<bool> {
+        if let Some(table) = self
+            .catalog_manager
+            .table(catalog_name, schema_name, table_name)
+            .await
+            .context(error::AccessCatalogSnafu)?
+        {
+            for r in region_numbers {
+                let region_exist =
+                    table
+                        .contains_region(*r)
+                        .await
+                        .with_context(|_| error::CheckRegionSnafu {
+                            table_name: format_full_table_name(
+                                catalog_name,
+                                schema_name,
+                                table_name,
+                            ),
+                            region_number: *r,
+                        })?;
+                if region_exist {
+                    return Ok(false);
+                }
+            }
+        }
+        // Returns true if table not exist
+        Ok(true)
+    }
+
+    async fn close_region_inner(
+        &self,
+        engine: String,
+        table_id: u32,
+        table_ref: &TableReference<'_>,
+        region_numbers: Vec<RegionNumber>,
+    ) -> Result<bool> {
+        let engine =
+            self.table_engine_manager
+                .engine(&engine)
+                .context(error::TableEngineNotFoundSnafu {
+                    engine_name: &engine,
+                })?;
+        let ctx = EngineContext::default();
+
+        if self
+            .check_table(
+                table_ref.catalog,
+                table_ref.schema,
+                table_ref.table,
+                &region_numbers,
+            )
+            .await?
+        {
+            return Ok(true);
+        }
+
+        if let Some(table) =
+            engine
+                .get_table(&ctx, table_ref)
+                .with_context(|_| error::GetTableSnafu {
+                    table_name: table_ref.to_string(),
+                })?
+        {
+            return if engine
+                .close_table(
+                    &ctx,
+                    CloseTableRequest {
+                        catalog_name: table_ref.catalog.to_string(),
+                        schema_name: table_ref.schema.to_string(),
+                        table_name: table_ref.table.to_string(),
+                        region_numbers: region_numbers.clone(),
+                    },
+                )
+                .await
+                .with_context(|_| error::CloseTableSnafu {
+                    table_name: table_ref.to_string(),
+                    region_numbers: region_numbers.clone(),
+                })? {
+                // Deregister table if The table released.
+                self.deregister_table(table_ref).await
+            } else {
+                // Registers table (update)
+                self.register_table(table_ref, table_id, table).await
+            };
+        }
+
+        // Table doesn't exist
+        Ok(true)
```

Comment:
I think it's better to return error if table not found. Normally under race condition would a table cannot be found here. Throw an error could indicate the caller that we have some concurrency issues to resolve.

Same to the open region case.